com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JdtUtils.java

```diff
@@ -78,4 +97,165 @@ public static IJavaProject getJavaProject(String projectName) {
         IProject project = root.getProject(projectName);
         return getJavaProject(project);
     }
+
+    /**
+     * Given the project name, return the corresponding project object.
+     * If the project doesn't exist, return null.
+     */
+    public static IProject getProject(String projectName) {
+        if (projectName == null) {
+            return null;
+        }
+        IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+        return root.getProject(projectName);
+    }
+
+    /**
+     * Compute the possible source containers that the specified project could be associated with.
+     * <p>
+     * If the project name is specified, it will put the source containers parsed from the specified project's
+     * classpath entries in the front of the result, then the other projects at the same workspace.
+     * </p>
+     * <p>
+     * Otherwise, just loop every projects at the current workspace and combine the parsed source containers directly.
+     * </p>
+     * @param projectName
+     *                  the project name.
+     * @return the possible source container list.
+     */
+    public static ISourceContainer[] getSourceContainers(String projectName) {
+        Set<ISourceContainer> containers = new LinkedHashSet<>();
+        List<IProject> projects = new ArrayList<>();
+
+        IProject targetProject = JdtUtils.getProject(projectName);
+        if (targetProject != null) {
+            projects.add(targetProject);
+        }
+
+        List<IProject> workspaceProjects = Arrays.asList(ResourcesPlugin.getWorkspace().getRoot().getProjects());
+        projects.addAll(workspaceProjects);
+
+        Set<IRuntimeClasspathEntry> calculated = new LinkedHashSet<>();
+
+        projects.stream().map(project -> JdtUtils.getJavaProject(project))
```

Comment:
Suggest using distinct() to avoid duplicates.

Reply:
fixed.